### PR TITLE
Query: Fixing extended functions in distributed querier

### PR DESCRIPTION
**Pull request by pedro-stanaka**
* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

- **Adding repro case for broken query with distributed engine**
- **Fixing problem with distributed queries and xfunctions**
- **Adding support for extended functions in tenancy enforcement**
- **Moving custom parser to new package**

## Verification

Added new acceptance tests and tested locally with two queriers.

### CODE REVIEW

**Comment by MichaHoffmann** [comment]
LGTM, can you separate the prometheus import from stdlib imports though please?

File: pkg/api/query/v1.go [hunk 1]
```diff
@@ -127,7 +127,15 @@ func (f *QueryEngineFactory) GetThanosEngine() ThanosEngine {
 		if f.remoteEngineEndpoints == nil {
 			f.thanosEngine = engine.New(engine.Opts{EngineOpts: f.engineOpts, Engine: f.GetPrometheusEngine(), EnableAnalysis: true, EnableXFunctions: f.enableXFunctions})
 		} else {
-			f.thanosEngine = engine.NewDistributedEngine(engine.Opts{EngineOpts: f.engineOpts, Engine: f.GetPrometheusEngine(), EnableAnalysis: true}, f.remoteEngineEndpoints)
+			f.thanosEngine = engine.NewDistributedEngine(
+				engine.Opts{
```
Comment (fpetkovski) [inline pkg/api/query/v1.go:131]: Looks like we want to define opts outside of the if statement and pass it into both engines.

Comment (pedro-stanaka) [reply pkg/api/query/v1.go]: done

**Comment by pedro-stanaka** [comment]
> LGTM, can you separate the prometheus import from stdlib imports though please?

@MichaHoffmann can you recheck?

**Review by fpetkovski — DISMISSED** [review]
Nice work

**Comment by pedro-stanaka** [comment]
> Should we forbid using `parser.ParseExpr` directly through faillint? What about other places where `parser.ParseExpr` is used? 🤔

I have a follow up PR where I will clean up all other places where we use `parser.ParserExpr` did not want to include on this one because I will touch more than 30 files. I can also create this fail lint rule if it is the case.

File: pkg/tenancy/tenancy_test.go [hunk 2]
```diff
@@ -196,3 +195,54 @@ func TestTenantProxyPassing(t *testing.T) {
 		_ = q.Series(&storepb.SeriesRequest{Matchers: seriesMatchers}, &storeSeriesServer{ctx: ctx})
 	})
 }
+
+func TestGetTenantFromHTTP(t *testing.T) {
```
Comment (GiedriusS) [inline pkg/tenancy/tenancy_test.go:199]: This test seems unrelated?

Comment (pedro-stanaka) [reply pkg/tenancy/tenancy_test.go]: removed.

File: pkg/cacheutil/memcached_client_test.go
Comment (MichaHoffmann) [inline pkg/cacheutil/memcached_client_test.go:1]: Ah, this has been flaky for a while I think!

**Review by MichaHoffmann — APPROVED** [review]
Lgtm

File: pkg/store/bucket_test.go
Comment (MichaHoffmann) [inline pkg/store/bucket_test.go:1]: That one is on me, thanks for fixing
